### DISCUSSION TITLE
trigger WithLatestFrom using a stable signal threshold crossing

### DISCUSSION THREAD

**Discussion by neurologic** [bonsai-rx/bonsai#1011] [Q&A]
I am getting an error, but this is probably not an issue with the program... Rather a hopeful ping for help on why I can't resolve the error message after spending a long time trying to figure out a solution from documentation, examples, issues page, and discussion page. Let me know if you prefer this in the Discussions section instead of Issues. Since I am generating an error, I thought others might search the Issues page if they encountered it in the future as well. 

**"error" I am stuck on**
"Unsupported format or combination of formatsin function cv::threshold, thresh.cpp(783)".

**ultimate goal**
To use one DAQmx>AnalogInput channel to trigger the visualization of another DAQmx>AnalogInput channel. (Data from both channels for offline analysis will be saved continuously as a binary file... which I already have working).

**what I have working so far**
I have part of the workflow doing what I want. I can get a triggered visualization of a specific channel from an **AnalogInput** node based off of a keyboard key press (I used an example from the docs on triggered audio playback to figure this out). 
![key-trigger](https://user-images.githubusercontent.com/4777333/184430400-751778e2-ffca-4b2d-830f-069921215b5d.PNG)
I use **SelectChannels** to subscribe to a specific input channel from a Nidaq ADC. I combine **KeyDown** and **SelectChannels** using **WithLatestFrom** and then use the **MatVisualizer** on **Item2** of the **WithLatestFrom**. 

**next step I am trying**
Using a stable threshold crossing from one **AnalogInput** channel to trigger the visualization of another **AnalogInput** channel. I thought I should do **AnalogInput**>**SelectChannels**>**Threshold(Dsp)** and then connect the output of **Threshold** to the **WithLatestFrom** in place of the **KeyDown** from my working workflow. I can't get past the Threshold step (I think once I do, I would use **GreaterThan** to get a Bool everytime the threshold output *stabley* increased from 0 to 1).

**THRESHOLD error**
![threshold-trigger_error](https://user-images.githubusercontent.com/4777333/184430412-5544313b-8e5c-422c-a51b-7eb3f8d6ddb4.PNG)
I cannot get past the error: "Unsupported format or combination of formatsin function cv::threshold, thresh.cpp(783)".
From the documentation I have found I am thinking that it is a mismatch between the type of data that the **Threshold** node expects and the type of data that the **SelectChannels** provides (OpenCV.Net.Mat). But I cannot find anywhere yet what the **Threshold** node expects. I have also tried thresholding MouseMove>X(int) output and that data type does not work either. I have tried converting the SelectChannels output to different datatypes but it has been unsuccessful trial and error so far. 
I have found lots of vision thresholding examples that have not seemed to apply to this type of signal. 

**DETECT SPIKES error**
I have also tried the **DetectSpikes** and get the same error... BUT if the **DetectSpikes** node is disabled, I **CAN** click the _Threshold_ dropdown menu (after starting the workflow) and see a live view of the data streaming in with a threshold line correctly reflecting the threshold I chose (and i can select others with that GUI). I am confused why I would be able to see this through the **DetectSpikes** node when it is disabled if the data type from **AnalogInput** was incompatible. 

oof.
Thank you in advance for anyone able to help me out. I am so grateful to have this tool available for use in the lab class that I teach.

**Comment by glopesdev** [comment]
@neurologic I'm sorry to hear your struggles with this. Unfortunately, the issue is actually a limitation of the underlying OpenCV DSP library threshold function. Strangely, `cvThreshold` does not support neither 16-bit unsigned nor 64-bit floating-point matrices as input, which is why you are ultimately seeing this message.

To work around this you will have to use the `ConvertScale` operator to first convert the signal to 32-bit floating-point, and then apply the threshold, like so:

<img src="https://user-images.githubusercontent.com/5315880/185754955-eb2159c8-73cc-480c-ad93-efe12e644815.png" height="50px">

The reason why the `DetectSpikes` editor worked is that the drawing of the threshold line is just a visualization, it is not actually applying the threshold, but simply setting the value of a property. If the node is disabled, no thresholding is really done.

Hope this helps.

**Reply by neurologic** [reply]
Yes that helps! Thank you so much.
The threshold and detect spike nodes are now working on the signals from Analog Input after converting to F32.
